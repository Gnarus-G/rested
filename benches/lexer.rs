use criterion::{criterion_group, criterion_main, Criterion};
use rested::lexer;

const SOURCE: &str = r#"
set BASE_URL env("base")

[env("token")]

let string = "death"

post /todos {
   header "Authorization" env("token")
   body {
    userId: 1,
    id: 999,
    "title": `delectus aut ${string}`,
    "completed": env("base")
  }
}
"#;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("tokenize", |b| {
        b.iter(|| {
            let lexer = lexer::Lexer::new(SOURCE);
            let _: Vec<_> = lexer.into_iter().collect();
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
