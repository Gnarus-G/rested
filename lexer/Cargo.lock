# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "base64"
version = "0.13.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9e1b586273c5702936fe7b7d6896644d8be71e6314cfe09d3167c95f712589e8"

[[package]]
name = "bitflags"
version = "1.3.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "bef38d45163c2f1dde094a7dfd33ccf595c92905c8f8f4fdc18d06fb1037718a"

[[package]]
name = "console"
version = "0.15.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c3d79fbe8970a77e3e34151cc13d3b3e248aa0faaecb9f6091fa07ebefe5ad60"
dependencies = [
 "encode_unicode",
 "lazy_static",
 "libc",
 "windows-sys",
]

[[package]]
name = "encode_unicode"
version = "0.3.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a357d28ed41a50f9c765dbfe56cbc04a64e53e5fc58ba79fbc34c10ef3df831f"

[[package]]
name = "insta"
version = "1.29.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9a28d25139df397cbca21408bb742cf6837e04cdbebf1b07b760caf971d6a972"
dependencies = [
 "console",
 "lazy_static",
 "linked-hash-map",
 "ron",
 "serde",
 "similar",
 "yaml-rust",
]

[[package]]
name = "lazy_static"
version = "1.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e2abad23fbc42b3700f2f279844dc832adb2b2eb069b2df918f455c4e18cc646"

[[package]]
name = "lexer"
version = "0.1.0"
dependencies = [
 "insta",
 "serde",
]

[[package]]
name = "libc"
version = "0.2.143"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "edc207893e85c5d6be840e969b496b53d94cec8be2d501b214f50daa97fa8024"

[[package]]
name = "linked-hash-map"
version = "0.5.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "0717cef1bc8b636c6e1c1bbdefc09e6322da8a9321966e8928ef80d20f7f770f"

[[package]]
name = "proc-macro2"
version = "1.0.56"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "2b63bdb0cd06f1f4dedf69b254734f9b45af66e4a031e42a7480257d9898b435"
dependencies = [
 "unicode-ident",
]

[[package]]
name = "quote"
version = "1.0.26"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "4424af4bf778aae2051a77b60283332f386554255d722233d09fbfc7e30da2fc"
dependencies = [
 "proc-macro2",
]

[[package]]
name = "ron"
version = "0.7.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "88073939a61e5b7680558e6be56b419e208420c2adb92be54921fa6b72283f1a"
dependencies = [
 "base64",
 "bitflags",
 "serde",
]

[[package]]
name = "serde"
version = "1.0.162"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "71b2f6e1ab5c2b98c05f0f35b236b22e8df7ead6ffbf51d7808da7f8817e7ab6"
dependencies = [
 "serde_derive",
]

[[package]]
name = "serde_derive"
version = "1.0.162"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a2a0814352fd64b58489904a44ea8d90cb1a91dcb6b4f5ebabc32c8318e93cb6"
dependencies = [
 "proc-macro2",
 "quote",
 "syn",
]

[[package]]
name = "similar"
version = "2.2.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "420acb44afdae038210c99e69aae24109f32f15500aa708e81d46c9f29d55fcf"

[[package]]
name = "syn"
version = "2.0.15"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a34fcf3e8b60f57e6a14301a2e916d323af98b0ea63c599441eec8558660c822"
dependencies = [
 "proc-macro2",
 "quote",
 "unicode-ident",
]

[[package]]
name = "unicode-ident"
version = "1.0.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e5464a87b239f13a63a501f2701565754bae92d243d4bb7eb12f6d57d2269bf4"

[[package]]
name = "windows-sys"
version = "0.42.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5a3e1820f08b8513f676f7ab6c1f99ff312fb97b553d30ff4dd86f9f15728aa7"
dependencies = [
 "windows_aarch64_gnullvm",
 "windows_aarch64_msvc",
 "windows_i686_gnu",
 "windows_i686_msvc",
 "windows_x86_64_gnu",
 "windows_x86_64_gnullvm",
 "windows_x86_64_msvc",
]

[[package]]
name = "windows_aarch64_gnullvm"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "597a5118570b68bc08d8d59125332c54f1ba9d9adeedeef5b99b02ba2b0698f8"

[[package]]
name = "windows_aarch64_msvc"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e08e8864a60f06ef0d0ff4ba04124db8b0fb3be5776a5cd47641e942e58c4d43"

[[package]]
name = "windows_i686_gnu"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c61d927d8da41da96a81f029489353e68739737d3beca43145c8afec9a31a84f"

[[package]]
name = "windows_i686_msvc"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "44d840b6ec649f480a41c8d80f9c65108b92d89345dd94027bfe06ac444d1060"

[[package]]
name = "windows_x86_64_gnu"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8de912b8b8feb55c064867cf047dda097f92d51efad5b491dfb98f6bbb70cb36"

[[package]]
name = "windows_x86_64_gnullvm"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "26d41b46a36d453748aedef1486d5c7a85db22e56aff34643984ea85514e94a3"

[[package]]
name = "windows_x86_64_msvc"
version = "0.42.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9aec5da331524158c6d1a4ac0ab1541149c0b9505fde06423b02f5ef0106b9f0"

[[package]]
name = "yaml-rust"
version = "0.4.5"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "56c1936c4cc7a1c9ab21a1ebb602eb942ba868cbd44a99cb7cdc5892335e1c85"
dependencies = [
 "linked-hash-map",
]
