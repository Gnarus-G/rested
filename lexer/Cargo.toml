[package]
name = "lexer"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1.0.159", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.29.0", features = ["ron"] }

[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
