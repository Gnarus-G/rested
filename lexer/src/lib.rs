pub mod locations;

use locations::Location;

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TokenKind {
    // keywords
    Get,
    Post,
    Put,
    Patch,
    Delete,
    Header,
    Body,
    Set,
    Let,

    Ident,

    // literals
    StringLiteral,
    MultiLineStringLiteral,
    Url,
    Pathname,

    Linecomment,
    Shebang,

    // operators
    Assign,

    // special characters
    DollarSignLBracket,
    LParen,
    RParen,
    LBracket,
    RBracket,
    LSquare,
    RSquare,
    Colon,
    AttributePrefix,
    Comma,
    End,

    //edge cases
    UnfinishedStringLiteral,
    UnfinishedMultiLineStringLiteral,
    IllegalToken,
}

#[derive(PartialEq)]
pub struct Token<'t> {
    pub kind: TokenKind,
    pub text: &'t str,
    pub start: Location,
}

impl<'source> Token<'source> {
    pub fn end_location(&self) -> Location {
        Location {
            line: self.start.line,
            col: self.start.col + self.text.len(),
        }
    }
}

impl<'i> std::fmt::Debug for Token<'i> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}({:?}) at {:?}", self.kind, self.text, self.start)
    }
}

trait CharaterTest {
    fn passes<P: Fn(&u8) -> bool>(&self, predicate: P) -> bool;
    fn is(&self, ch: u8) -> bool {
        self.passes(|&c| ch == c)
    }
}

impl CharaterTest for Option<&u8> {
    fn passes<P: Fn(&u8) -> bool>(&self, predicate: P) -> bool {
        match self {
            Some(c) => predicate(c),
            None => false,
        }
    }
}

#[derive(Debug)]
pub struct Lexer<'i> {
    input: &'i [u8],
    position: usize,
    cursor: Location,
    inside_multiline_string: bool,
}

impl<'i> Lexer<'i> {
    pub fn new(input: &'i str) -> Self {
        Self {
            input: input.as_bytes(),
            position: 0,
            cursor: Location { line: 0, col: 0 },
            inside_multiline_string: false,
        }
    }

    pub fn input(&self) -> &'i str {
        std::str::from_utf8(&self.input).expect("input should only contain utf-8 characters")
    }

    fn input_slice(&self, range: impl std::slice::SliceIndex<[u8], Output = [u8]>) -> &'i str {
        std::str::from_utf8(&self.input[range]).expect("input should only contain utf-8 characters")
    }

    fn char_at(&self, position: usize) -> Option<&u8> {
        if position < self.input.len() {
            return Some(&self.input[position]);
        }
        return None;
    }

    fn ch(&self) -> Option<&u8> {
        self.char_at(self.position)
    }

    fn step(&mut self) {
        if self.peek_char().is_none() {
            return self.position += 1;
        }
        self.check_and_bump_new_line();
        self.position += 1;
    }

    fn check_and_bump_new_line(&mut self) {
        if let Some(b'\n') = self.ch() {
            self.cursor.line += 1;
            self.cursor.col = 0;
        } else {
            self.cursor.col += 1;
        };
    }

    fn peek_n_char(&self, n: usize) -> Option<&u8> {
        self.char_at(self.position + n)
    }

    fn peek_char(&self) -> Option<&u8> {
        self.peek_n_char(1)
    }

    /// Assumes that the character at the current position, immediately before calling
    /// this function is also true for the predicate function given.
    fn read_while<P: Fn(&u8) -> bool>(&mut self, predicate: P) -> (usize, usize) {
        let start_pos = self.position;

        while self.peek_char().passes(&predicate) {
            self.step();
        }

        return (start_pos, self.position + 1);
    }

    fn skip_whitespace(&mut self) {
        while self.ch().passes(|c| c.is_ascii_whitespace()) {
            self.step();
        }
    }

    pub fn next(&mut self) -> Token<'i> {
        use TokenKind::*;

        self.skip_whitespace();

        let ch = match self.ch() {
            Some(ch) => ch,
            None => {
                return Token {
                    kind: TokenKind::End,
                    start: self.cursor,
                    text: "",
                }
            }
        };

        let t = match ch {
            b'"' if self.peek_char().is(b'"') => self.empty_string_literal(),
            b'"' => self.string_literal(),
            b'`' if self.peek_char().is(b'`') => self.empty_string_literal(),
            b'`' => self.multiline_string_literal(),
            b'$' if self.peek_char().is(b'{') => {
                let token = Token {
                    kind: DollarSignLBracket,
                    text: "${",
                    start: self.cursor,
                };
                self.step();
                token
            }
            b'}' if self.peek_char().is(b'`') && self.inside_multiline_string => {
                self.step(); // eat the curly
                self.step(); // eat the backtick
                self.inside_multiline_string = false;
                self.next()
            }
            b'}' if self.inside_multiline_string => {
                self.step(); // eat the curly
                self.multiline_string_literal()
            }
            b'(' => Token {
                kind: LParen,
                start: self.cursor,
                text: "(",
            },
            b')' => Token {
                kind: RParen,
                start: self.cursor,
                text: ")",
            },
            b'{' => Token {
                kind: LBracket,
                start: self.cursor,
                text: "{",
            },
            b'}' => Token {
                kind: RBracket,
                start: self.cursor,
                text: "}",
            },
            b'[' => Token {
                kind: LSquare,
                start: self.cursor,
                text: "[",
            },
            b']' => Token {
                kind: RSquare,
                start: self.cursor,
                text: "]",
            },
            b',' => Token {
                kind: Comma,
                text: ",",
                start: self.cursor,
            },
            b':' => Token {
                kind: Colon,
                text: ":",
                start: self.cursor,
            },
            b'=' => Token {
                kind: Assign,
                text: "=",
                start: self.cursor,
            },
            b'@' => Token {
                kind: AttributePrefix,
                text: "@",
                start: self.cursor,
            },
            b'/' if self.peek_char().is(b'/') => self.line_comment(),
            b'/' => self.pathname(),
            b'#' if self.peek_char().is(b'!') => self.shebang(),
            c if c.is_ascii_alphabetic() => self.keyword_or_identifier(),
            _ => Token {
                kind: IllegalToken,
                text: std::str::from_utf8(&self.input[self.position..self.position + 1]).unwrap(),
                start: self.cursor,
            },
        };

        self.step();

        t
    }

    fn multiline_string_literal(&mut self) -> Token<'i> {
        self.inside_multiline_string = true;
        let location = self.cursor;
        let start_pos = self.position;

        self.step();

        let (s, e) = loop {
            let is_at_end = self.ch().is(b'`');
            let dollar_curly_ahead = self.peek_char().is(b'$') && self.peek_n_char(2).is(b'{');

            if dollar_curly_ahead {
                break (start_pos, self.position + 1);
            }

            if is_at_end {
                self.inside_multiline_string = false;
                break (start_pos, self.position + 1);
            }

            if let None = self.ch() {
                return Token {
                    kind: TokenKind::UnfinishedMultiLineStringLiteral,
                    start: location,
                    text: self.input_slice(start_pos..self.position),
                };
            }

            self.step();
        };

        let string = self.input_slice(s..e);

        Token {
            kind: TokenKind::MultiLineStringLiteral,
            start: location,
            text: string,
        }
    }

    fn string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;

        let (s, e) = self.read_while(|&c| c != b'"' && c != b'\n');

        match self.peek_char() {
            Some(b'\n') | None => {
                return Token {
                    kind: TokenKind::UnfinishedStringLiteral,
                    start: location,
                    text: self.input_slice(..e),
                };
            }
            _ => {}
        }

        let string = self.input_slice(s..=e);

        self.step();

        Token {
            kind: TokenKind::StringLiteral,
            start: location,
            text: string,
        }
    }

    fn empty_string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;
        self.step();
        Token {
            kind: TokenKind::StringLiteral,
            start: location,
            text: "",
        }
    }

    fn keyword_or_identifier(&mut self) -> Token<'i> {
        let location = self.cursor;
        let (s, e) = self.read_while(|&c| c.is_ascii_alphabetic() || c == b'_');
        let string = self.input_slice(s..e);

        use TokenKind::*;

        match string {
            "let" => Token {
                kind: Let,
                start: location,
                text: string,
            },
            "get" => Token {
                kind: Get,
                start: location,
                text: string,
            },
            "post" => Token {
                kind: Post,
                start: location,
                text: string,
            },
            "put" => Token {
                kind: Put,
                start: location,
                text: string,
            },
            "patch" => Token {
                kind: Patch,
                start: location,
                text: string,
            },
            "delete" => Token {
                kind: Delete,
                start: location,
                text: string,
            },
            "header" => Token {
                kind: Header,
                start: location,
                text: string,
            },
            "set" => Token {
                kind: Set,
                start: location,
                text: string,
            },
            "body" => Token {
                kind: Body,
                start: location,
                text: string,
            },
            "http" | "https" => {
                let (.., e) = self.read_while(|&c| !c.is_ascii_whitespace());
                let s = self.input_slice(s..e);
                Token {
                    kind: Url,
                    start: location,
                    text: s,
                }
            }
            s => Token {
                kind: Ident,
                start: location,
                text: s,
            },
        }
    }

    fn pathname(&mut self) -> Token<'i> {
        let location = self.cursor;
        let (s, e) = self.read_while(|&c| !c.is_ascii_whitespace());
        let string = self.input_slice(s..e);

        Token {
            kind: TokenKind::Pathname,
            start: location,
            text: string,
        }
    }

    fn shebang(&mut self) -> Token<'i> {
        let location = self.cursor;
        let (s, e) = self.read_while(|&c| c != b'\n');
        let string = self.input_slice(s..e);
        Token {
            text: string,
            start: location,
            kind: TokenKind::Shebang,
        }
    }

    fn line_comment(&mut self) -> Token<'i> {
        let location = self.cursor;
        let (s, e) = self.read_while(|&c| c != b'\n');
        let string = self.input_slice(s..e);
        Token {
            kind: TokenKind::Linecomment,
            text: string,
            start: location,
        }
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = Token<'source>;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.next();

        if let TokenKind::End = token.kind {
            return None;
        }

        Some(token)
    }
}
