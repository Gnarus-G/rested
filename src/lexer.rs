use std::ops::Range;

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum TokenKind {
    // keywords
    Get,
    Post,
    Header,
    Body,

    Ident,

    // literals
    StringLiteral,
    MultiLineStringLiteral,
    Url,

    // special characters
    LParen,
    RParen,
    LBracket,
    RBracket,
    End,

    //edge cases
    UnfinishedStringLiteral,
    UnfinishedMultiLineStringLiteral,
    IllegalToken,
}

#[derive(PartialEq, Clone, Copy)]
pub struct Location {
    pub line: usize,
    pub col: usize,
}

impl std::fmt::Debug for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.line, self.col)
    }
}

#[derive(PartialEq)]
pub struct Token<'t> {
    pub kind: TokenKind,
    pub text: &'t str,
    pub location: Location,
}

impl<'i> std::fmt::Debug for Token<'i> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}({:?}) at {:?}", self.kind, self.text, self.location)
    }
}

trait CharaterTest {
    fn passes<P: Fn(&u8) -> bool>(&self, predicate: P) -> bool;
    fn is(&self, ch: u8) -> bool {
        self.passes(|&c| ch == c)
    }
}

impl CharaterTest for Option<&u8> {
    fn passes<P: Fn(&u8) -> bool>(&self, predicate: P) -> bool {
        match self {
            Some(c) => predicate(c),
            None => false,
        }
    }
}

#[derive(Debug)]
pub struct Lexer<'i> {
    input: &'i [u8],
    position: usize,
    cursor: Location,
}

impl<'i> Lexer<'i> {
    pub fn new(input: &'i str) -> Self {
        Self {
            input: input.as_bytes(),
            position: 0,
            cursor: Location { line: 0, col: 0 },
        }
    }

    pub fn input(&self) -> &'i str {
        std::str::from_utf8(&self.input).expect("input should only contain utf-8 characters")
    }

    fn input_slice(&self, range: Range<usize>) -> &'i str {
        std::str::from_utf8(&self.input[range]).expect("input should only contain utf-8 characters")
    }

    fn char_at(&self, position: usize) -> Option<&u8> {
        if position < self.input.len() {
            return Some(&self.input[position]);
        }
        return None;
    }

    fn ch(&self) -> Option<&u8> {
        self.char_at(self.position)
    }

    fn step(&mut self) {
        if let Some(b'\n') = self.ch() {
            self.cursor.line += 1;
            self.cursor.col = 0;
        } else {
            self.cursor.col += 1;
        };

        self.position += 1;
    }

    fn peek_char(&self) -> Option<&u8> {
        self.char_at(self.position + 1)
    }

    /// Assumes that the character at the current position, immediately before calling
    /// this function is also true for the predicate function given.
    fn read_while<P: Fn(&u8) -> bool>(&mut self, predicate: P) -> (usize, usize) {
        let start_pos = self.position;

        while self.peek_char().passes(&predicate) {
            self.step();
        }

        return (start_pos, self.position + 1);
    }

    fn skip_whitespace(&mut self) {
        while self.ch().passes(|c| c.is_ascii_whitespace()) {
            self.step();
        }
    }

    pub fn next(&mut self) -> Token<'i> {
        use TokenKind::*;

        self.skip_whitespace();

        let ch = match self.ch() {
            Some(ch) => ch,
            None => {
                return Token {
                    kind: TokenKind::End,
                    location: self.cursor,
                    text: "",
                }
            }
        };

        let t = match ch {
            b'"' if self.peek_char().is(b'"') => self.empty_string_literal(),
            b'"' => self.string_literal(),
            b'`' if self.peek_char().is(b'`') => self.empty_string_literal(),
            b'`' => self.multiline_string_literal(),
            b'(' => Token {
                kind: LParen,
                location: self.cursor,
                text: "(",
            },
            b')' => Token {
                kind: RParen,
                location: self.cursor,
                text: ")",
            },
            b'{' => Token {
                kind: LBracket,
                location: self.cursor,
                text: "{",
            },
            b'}' => Token {
                kind: RBracket,
                location: self.cursor,
                text: "}",
            },
            c if c.is_ascii_alphabetic() => self.keyword_or_identifier(),
            _ => Token {
                kind: IllegalToken,
                text: std::str::from_utf8(&self.input[self.position..self.position + 1]).unwrap(),
                location: self.cursor,
            },
        };

        self.step();

        t
    }

    fn multiline_string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;

        self.step(); //eat the opening quote

        let (s, e) = self.read_while(|&c| c != b'`');
        let string = self.input_slice(s..e);

        self.step(); //eat the closing quote

        if let None = self.ch() {
            return Token {
                kind: TokenKind::UnfinishedMultiLineStringLiteral,
                location,
                text: string,
            };
        }

        Token {
            kind: TokenKind::MultiLineStringLiteral,
            location,
            text: string,
        }
    }

    fn string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;

        self.step(); //eat the opening quote

        let (s, e) = self.read_while(|&c| c != b'"' && c != b'\n');
        let string = self.input_slice(s..e);

        self.step(); //eat the closing quote or newline character

        match self.ch() {
            Some(b'\n') | None => {
                return Token {
                    kind: TokenKind::UnfinishedStringLiteral,
                    location,
                    text: string,
                };
            }
            _ => {}
        }

        Token {
            kind: TokenKind::StringLiteral,
            location,
            text: string,
        }
    }

    fn empty_string_literal(&mut self) -> Token<'i> {
        let location = self.cursor;
        self.step();
        Token {
            kind: TokenKind::StringLiteral,
            location,
            text: "",
        }
    }

    fn keyword_or_identifier(&mut self) -> Token<'i> {
        let location = self.cursor;
        let (s, e) = self.read_while(|c| c.is_ascii_alphabetic());
        let string = self.input_slice(s..e);

        use TokenKind::*;

        match string {
            "get" => Token {
                kind: Get,
                location,
                text: string,
            },
            "post" => Token {
                kind: Post,
                location,
                text: string,
            },
            "header" => Token {
                kind: Header,
                location,
                text: string,
            },
            "body" => Token {
                kind: Body,
                location,
                text: string,
            },
            "http" | "https" => {
                let (.., e) = self.read_while(|&c| !c.is_ascii_whitespace());
                let s = self.input_slice(s..e);
                Token {
                    kind: Url,
                    location,
                    text: s,
                }
            }
            s => Token {
                kind: Ident,
                location,
                text: s,
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use TokenKind::*;

    impl Into<Location> for (usize, usize) {
        fn into(self) -> Location {
            Location {
                line: self.0,
                col: self.1,
            }
        }
    }

    macro_rules! assert_lexes {
        ($input:literal, $tokens:expr) => {
            let mut lexer = Lexer::new($input);

            for token in $tokens {
                assert_eq!(lexer.next(), token);
            }
        };
    }

    #[test]
    fn lex_string_literals() {
        assert_lexes!(
            r#""hello""#,
            [Token {
                kind: StringLiteral,
                text: "hello",
                location: (0, 0).into()
            },]
        );

        assert_lexes!(
            r#""hello"#,
            [Token {
                kind: UnfinishedStringLiteral,
                text: "hello",
                location: (0, 0).into()
            },]
        );

        assert_lexes!(
            r#"
"hello
"world
"#,
            [
                Token {
                    kind: UnfinishedStringLiteral,
                    text: "hello",
                    location: (1, 0).into()
                },
                Token {
                    kind: UnfinishedStringLiteral,
                    text: "world",
                    location: (2, 0).into()
                }
            ]
        );

        assert_lexes!(
            r#" "" "" ``"#,
            [
                Token {
                    kind: StringLiteral,
                    text: "",
                    location: (0, 1).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "",
                    location: (0, 4).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "",
                    location: (0, 7).into()
                }
            ]
        );

        assert_lexes!(
            r#" { "Bearer token" } "#,
            [
                Token {
                    kind: LBracket,
                    text: "{",
                    location: (0, 1).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "Bearer token",
                    location: (0, 3).into()
                },
                Token {
                    kind: RBracket,
                    text: "}",
                    location: (0, 18).into()
                }
            ]
        );

        assert_lexes!(
            r#"`
{
    stuff
}`

`
stuff"#,
            [
                Token {
                    kind: MultiLineStringLiteral,
                    text: "\n{\n    stuff\n}",
                    location: (0, 0).into()
                },
                Token {
                    kind: UnfinishedMultiLineStringLiteral,
                    text: "\nstuff",
                    location: (5, 0).into()
                }
            ]
        );
    }

    #[test]
    fn lex_get_url() {
        assert_lexes!(
            "get http://localhost",
            [
                Token {
                    kind: Get,
                    text: "get",
                    location: (0, 0).into()
                },
                Token {
                    kind: Url,
                    text: "http://localhost",
                    location: (0, 4).into(),
                }
            ]
        );
    }

    #[test]
    fn lex_get_url_with_header() {
        assert_lexes!(
            "get http://localhost { header \"Authorization\" \"Bearer token\" }",
            vec![
                Token {
                    kind: Get,
                    location: (0, 0).into(),
                    text: "get"
                },
                Token {
                    kind: Url,
                    location: (0, 4).into(),
                    text: "http://localhost"
                },
                Token {
                    kind: LBracket,
                    location: (0, 21).into(),
                    text: "{"
                },
                Token {
                    kind: Header,
                    location: (0, 23).into(),
                    text: "header"
                },
                Token {
                    kind: StringLiteral,
                    location: (0, 30).into(),
                    text: "Authorization"
                },
                Token {
                    kind: StringLiteral,
                    location: (0, 46).into(),
                    text: "Bearer token"
                },
                Token {
                    kind: RBracket,
                    location: (0, 61).into(),
                    text: "}"
                },
            ]
        );
    }

    #[test]
    fn lex_get_url_over_many_lines() {
        assert_lexes!(
            "get\nhttp://localhost",
            [
                Token {
                    kind: Get,
                    text: "get",
                    location: (0, 0).into()
                },
                Token {
                    kind: Url,
                    text: "http://localhost",
                    location: (1, 0).into(),
                }
            ]
        );

        assert_lexes!(
            r#"get 
    http://localhost 
{
}"#,
            [
                Token {
                    kind: Get,
                    text: "get",
                    location: (0, 0).into()
                },
                Token {
                    kind: Url,
                    text: "http://localhost",
                    location: (1, 4).into(),
                },
                Token {
                    kind: LBracket,
                    text: "{",
                    location: (2, 0).into(),
                },
                Token {
                    kind: RBracket,
                    text: "}",
                    location: (3, 0).into(),
                }
            ]
        );
    }

    #[test]
    fn lex_get_url_with_header_and_body() {
        assert_lexes!(
            r#"
post http://localhost { 
    header "Authorization" "Bearer token" 
    body "{neet: 1337}" 
}"#,
            vec![
                Token {
                    kind: Post,
                    location: (1, 0).into(),
                    text: "post"
                },
                Token {
                    kind: Url,
                    location: (1, 5).into(),
                    text: "http://localhost"
                },
                Token {
                    kind: LBracket,
                    location: (1, 22).into(),
                    text: "{"
                },
                Token {
                    kind: Header,
                    location: (2, 4).into(),
                    text: "header"
                },
                Token {
                    kind: StringLiteral,
                    location: (2, 11).into(),
                    text: "Authorization"
                },
                Token {
                    kind: StringLiteral,
                    location: (2, 27).into(),
                    text: "Bearer token"
                },
                Token {
                    kind: Body,
                    location: (3, 4).into(),
                    text: "body"
                },
                Token {
                    kind: StringLiteral,
                    location: (3, 9).into(),
                    text: "{neet: 1337}"
                },
                Token {
                    kind: RBracket,
                    location: (4, 0).into(),
                    text: "}"
                },
            ]
        );
    }

    #[test]
    fn lex_call_expression() {
        assert_lexes!(
            r#"env() env("stuff")"#,
            [
                Token {
                    kind: Ident,
                    text: "env",
                    location: (0, 0).into()
                },
                Token {
                    kind: LParen,
                    text: "(",
                    location: (0, 3).into()
                },
                Token {
                    kind: RParen,
                    text: ")",
                    location: (0, 4).into()
                },
                Token {
                    kind: Ident,
                    text: "env",
                    location: (0, 6).into()
                },
                Token {
                    kind: LParen,
                    text: "(",
                    location: (0, 9).into()
                },
                Token {
                    kind: StringLiteral,
                    text: "stuff",
                    location: (0, 10).into()
                },
                Token {
                    kind: RParen,
                    text: ")",
                    location: (0, 17).into()
                }
            ]
        );
    }
}
