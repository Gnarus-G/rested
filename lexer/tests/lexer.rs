use insta::assert_debug_snapshot;

macro_rules! assert_lexes {
    ($input:literal) => {
        let lexer = Lexer::new($input);

        insta::with_settings!({
             description => $input
        }, {
            assert_debug_snapshot!(lexer.into_iter().collect::<Vec<Token>>());
        })
    }
}

#[test]
fn lex_put_patch_delete() {
    assert_lexes!(
        r#"
put /api {}
patch /api {}
delete /api {}
"#
    );
}

use lexer::*;

#[test]
fn lex_string_literals() {
    assert_lexes!(r#""hello""#);

    assert_lexes!(r#""hello"#);

    assert_lexes!(
        r#"
"hello
"world
"#
    );

    assert_lexes!(r#" "" "" ``"#);

    assert_lexes!(r#" { "Bearer token" } "#);

    assert_lexes!(
        r#"`
{
    stuff
}`

`
stuff"#
    );
}

#[test]
fn lex_get_url() {
    assert_lexes!("get http://localhost");
}

#[test]
fn lex_get_url_with_header() {
    assert_lexes!("get http://localhost { header \"Authorization\" \"Bearer token\" }");
}

#[test]
fn lex_get_url_over_many_lines() {
    assert_lexes!("get\nhttp://localhost");

    assert_lexes!(
        r#"get 
    http://localhost 
{
}"#
    );
}

#[test]
fn lex_get_url_with_header_and_body() {
    assert_lexes!(
        r#"
post http://localhost { 
    header "Authorization" "Bearer token" 
    body "{neet: 1337}" 
}"#
    );
}

#[test]
fn lex_call_expression() {
    assert_lexes!(r#"env() env("stuff")"#);
}

#[test]
fn lex_template_literals() {
    assert_lexes!(r#"`stuff${"interpolated"}(things${env("dead_night")}` `dohickeys`"#);

    assert_lexes!(r#"`a${"temp"}` }}"#);
}
