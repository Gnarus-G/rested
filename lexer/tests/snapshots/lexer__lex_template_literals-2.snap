---
source: lexer/tests/lexer.rs
description: "`a${\"temp\"}` }}"
expression: "lexer.into_iter().collect::<Vec<Token>>()"
---
[
    MultiLineStringLiteral("`a") at 0:0,
    DollarSignLBracket("${") at 0:2,
    StringLiteral("\"temp\"") at 0:4,
    RBracket("}") at 0:13,
]
